### PR TITLE
feat(field-level-auth): Add field level auth support via the @auth directive

### PR DESCRIPTION
**Pull request by mikeparisstuff**
*Issue #, if available:*

#1043

*Description of changes:*

Adding support for field level authorization via the @auth directive.

This addresses proposals 1 & 4 in #1043.
    
    Changes:
    - Added the 'operations' argument to the AuthRule input used by the @auth directive.
    - Made backwards compatible changes such that @auth directives that use the 'operations' argument protect @connection resolvers.
    - Added support for field level @auth directives. Protect query resolvers on any type and mutations on @models.
    - Many new e2e tests for field level authorization checks.
    - Refactor existing tests to make debugging and coverage easier to understand.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by mikeparisstuff** [comment]
Implements proposals 1 & 5 from #1043.

File: packages/graphql-auth-transformer/src/__tests__/OperationsArgument.test.ts [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        ]
+    })
+    const out = transformer.transform(validSchema)
+    expect(out).toBeDefined()
+    expect(
+        out.rootStack.Resources[ResourceConstants.RESOURCES.GraphQLAPILogicalID].Properties.AuthenticationType
+    ).toEqual('AMAZON_COGNITO_USER_POOLS')
+    expect(
+        out.resolvers['Query.getPost.res.vtl']
+    ).toContain('Authorization rule:')
```
Comment (yuth) [inline packages/graphql-auth-transformer/src/__tests__/OperationsArgument.test.ts]: Do you think these would be good candidate for snapshot tests?

File: packages/graphql-transformers-e2e-tests/src/__tests__/ModelAuthTransformer.e2e.test.ts [hunk 2]
```diff
@@ -364,354 +374,2576 @@ afterAll(async () => {
     }
 })
 
+// /**
```
Comment (yuth) [inline packages/graphql-transformers-e2e-tests/src/__tests__/ModelAuthTransformer.e2e.test.ts]: Do we need these to be commented?

File: packages/graphql-transformers-e2e-tests/src/__tests__/ModelAuthTransformer.e2e.test.ts [hunk 3]
```diff
@@ -2812,5 +5044,8 @@ test(`Test createTestIdentity as admin.`, async () => {
     } catch (e) {
         console.error(e)
         expect(e).toBeUndefined();
+=======
```
Comment (yuth) [inline packages/graphql-transformers-e2e-tests/src/__tests__/ModelAuthTransformer.e2e.test.ts]: Merge conflicts

File: packages/graphql-auth-transformer/src/__tests__/OperationsArgument.test.ts (hunk 1 above)
Comment (mikeparisstuff) [reply packages/graphql-auth-transformer/src/__tests__/OperationsArgument.test.ts]: Perhaps. Do you mean storing the output and then diffing them between runs?

Comment (yuth) [reply packages/graphql-auth-transformer/src/__tests__/OperationsArgument.test.ts]: Yes. Jest has built in support for that. 
```typescript
expect(out.resolvers['Query.getPost.res.vtl']).toMatchSnapshot();
```
this should do it

**Comment by sebastienfi** [comment]
> Implements proposals 1 & 5 from #1043.

I think you meant "1 & 4".

**Pull request merged**